### PR TITLE
Reject connection related headers in HTTP/2 frames

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
In https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.2 is says "any message containing connection-specific header fields MUST be treated as malformed".
We already filter these out in our HTTP/1 to HTTP/2 translation code, but if the HTTP/2 frames are originally sent to us with these headers, then we must reject them as malformed.

Modification:
Improve the header validation code used by the DefaultHttp2HeadersDecoder to check for, and reject, any connection related headers.

Result:
Frames with illegal headers are now correctly rejected, though only when header validation is turned on.

### CODE REVIEW

File: codec-http2/src/test/resources/io/netty/handler/codec/http2/testdata/testStaticTableEntries.json [hunk 1]
```diff
@@ -62,7 +61,7 @@
         "8182 8384 8586 87 8f90",
         "9192 9394 9596 9798 999a 9b9c 9d9e 9fa0",
         "a1a2 a3a4 a5a6 a7a8 a9aa abac adae afb0",
-        "b1b2 b3b4 b5b6 b7b8 b9ba bbbc bd"
+        "b1b2 b3b4 b5b6 b7b8 babb bcbd"
```
Comment (normanmaurer) [inline codec-http2/src/test/resources/io/netty/handler/codec/http2/testdata/testStaticTableEntries.json:64]: why the changes in this file ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http2/src/test/resources/io/netty/handler/codec/http2/testdata/testStaticTableEntries.json (hunk 1 above)
Comment (chrisvest) [reply codec-http2/src/test/resources/io/netty/handler/codec/http2/testdata/testStaticTableEntries.json:64]: `transfer-encoding` is a connection related header.

Comment (violetagg) [reply codec-http2/src/test/resources/io/netty/handler/codec/http2/testdata/testStaticTableEntries.json:64]: @chrisvest @normanmaurer Here is the latest RFC and the changes related to headers https://datatracker.ietf.org/doc/html/rfc9113#appendix-B

Comment (chrisvest) [reply codec-http2/src/test/resources/io/netty/handler/codec/http2/testdata/testStaticTableEntries.json:64]: Which in turn references the list of connection headers from https://www.rfc-editor.org/rfc/rfc9110.html#name-connection

**Review by normanmaurer — APPROVED** [review]

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 2]
```diff
@@ -404,10 +408,50 @@ private static HeaderType validate(int streamId, CharSequence name,
...
+        // - upgrade (7 chars)
+        // - connection (10 chars)
+        // - keep-alive (10 chars)
+        // - proxy-connection (16 chars)
+        // - transfer-encoding (17 chars)
+        // - upgrade-insecure-requests (25 chars)
+        //
+        // We scan for these based on the length, then double-check any matching name.
+        int len = name.length();
+        if (len < 7 || len > 25) {
```
Comment (idelpivnitskiy) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:435]: Have you considered using `switch-case` to find exact lengths?

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoderTest.java [hunk 3]
```diff
@@ -129,9 +133,50 @@ public void execute() throws Throwable {
 
     @Test
     public void duplicatePseudoHeadersMustFailValidation() throws Exception {
-        final ByteBuf buf = encode(b(":authority"), b("abc"), b(":authority"), b("def"));
+        final DefaultHttp2HeadersDecoder decoder = new DefaultHttp2HeadersDecoder(true);
+        verifyValidationFails(decoder, encode(b(":authority"), b("abc"), b(":authority"), b("def")));
+    }
+
+    @Test
+    void decodingTrailersTeHeaderMustNotFailValidation() throws Exception {
```
Comment (idelpivnitskiy) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoderTest.java:141]: Would be nice to have a test that validates the remote peer receives a reset frame with `PROTOCOL_ERROR`

**Review by idelpivnitskiy — COMMENTED** [review]
Glad to see this in netty, thank you!

**Comment by chrisvest** [comment]
@idelpivnitskiy I'll address these comments in #12760

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 4]
```diff
@@ -404,10 +408,50 @@ private static HeaderType validate(int streamId, CharSequence name,
...
 
+    @SuppressWarnings("deprecation") // We need to check for deprecated headers as well.
+    private static boolean isConnectionHeader(CharSequence name) {
+        // These are the known standard and non-standard connection related headers:
+        // - upgrade (7 chars)
+        // - connection (10 chars)
+        // - keep-alive (10 chars)
+        // - proxy-connection (16 chars)
+        // - transfer-encoding (17 chars)
+        // - upgrade-insecure-requests (25 chars)
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:431]: It's defined in https://www.w3.org/TR/upgrade-insecure-requests/ and marked as not suitable for H/2 on https://en.wikipedia.org/wiki/List_of_HTTP_header_fields#Common_non-standard_request_fields

I'm not finding unambiguous wording that it's connection related or forbidden in H/2, though.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoderTest.java (hunk 3 above)
Comment (chrisvest) [reply codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoderTest.java:141]: Looks like we don't have any integration tests.

@normanmaurer Do you know of a good place to add a test like this, or is this an entirely new test?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:431]: Thank you!
I don't have any preference to either side for this header. Interesting why someone decided to mark it as "Must not be used with HTTP/2" on wikipedia. Maybe bcz all browsers implement HTTP/2 only over TLS.

**Comment by glennosss** [comment]
@chrisvest @normanmaurer @idelpivnitskiy This change is breaking Chrome and Firefox which both (rightly or wrongly) send the `upgrade-insecure-requests` header:

```
[DEBUG|2022-09-13 10:23:29.419|netty-worker1 ||io.netty.handler.codec.http2.Http2FrameCodec] Stream exception thrown for unknown stream 1.
io.netty.handler.codec.http2.Http2Exception$StreamException: Illegal connection-specific header 'upgrade-insecure-requests' encountered.
	at io.netty.handler.codec.http2.Http2Exception.streamError(Http2Exception.java:153)
	at io.netty.handler.codec.http2.HpackDecoder.validate(HpackDecoder.java:412)
	at io.netty.handler.codec.http2.HpackDecoder.access$000(HpackDecoder.java:56)
	at io.netty.handler.codec.http2.HpackDecoder$Http2HeadersSink.appendToHeaderList(HpackDecoder.java:639)
	at io.netty.handler.codec.http2.HpackDecoder.insertHeader(HpackDecoder.java:506)
	at io.netty.handler.codec.http2.HpackDecoder.decode(HpackDecoder.java:294)
	at io.netty.handler.codec.http2.HpackDecoder.decode(HpackDecoder.java:131)
```

With Netty 4.1.80 and below I was able to run with `validateHeaders(true)` (the default). However with Netty 4.1.81 the only work around I've found is;
```
        Http2FrameCodecBuilder builder = Http2FrameCodecBuilder.forServer();
        builder.validateHeaders(false);
```

I recommend people avoid updating to Netty 4.1.81 until this is resolved. I've had to downgrade to Netty 4.1.80.

**Comment by chrisvest** [comment]
@glennosss That's unfortunate. I'll take a look. Can you please make a separate issue to track this?

**Comment by glennosss** [comment]
> @glennosss That's unfortunate. I'll take a look. Can you please make a separate issue to track this?

Created! https://github.com/netty/netty/issues/12798
